src/Microsoft.ML.LightGbm/LightGbmMulticlassTrainer.cs

```diff
@@ -216,9 +216,20 @@ private protected override void CheckDataValid(IChannel ch, RoleMappedData data)
             }
         }
 
+        private protected override void InitializeBeforeTraining()
+        {
+            _numClass = -1;
+            _tlcNumClass = 0;
+
+            //MYTODO: Include more initializations, of TrainedEnsemble, for example?
+            //For example:
+            //TrainedEnsemble = null;
```

Comment:
So here I re-initialize the `_numClass` and `_tlcNumClass` fields.

But I wonder if there's any reason to also re-initialize any other field, such as `TrainedEnsemble`, since trainers are supposed to be stateless.

After looking into the code I don't find any reason to re-initialize any other variable... but I would still appreciate any feedback on this regard.

Reply:
I think you are right. `TrainedEnsemble` and `FeatureCount` get assigned new values anyway each time `Fit()` is called so they don't need to be initialized in the new method you are introducing. I looked at the other fields of `LightGbmTrainerBase`, and I have additional thoughts about the two fields `GbmOptions` and `ParallelTraining`. I will write them down in a separate comment.

---
In reply to: [362128849](https://github.com/dotnet/machinelearning/pull/4608#discussion_r362128849) [](ancestors = 362128849)